### PR TITLE
Reordering public input and outputs before private inputs

### PR DESCRIPTION
**Pull request by katat**
https://github.com/zksecurity/noname/issues/70

### CODE REVIEW

File: src/backends/r1cs/mod.rs [hunk 1]
```diff
@@ -310,6 +311,21 @@ where
...
+    fn init_circuit(&mut self, func: &mut FunctionDef) {
+        // reorder the arguments by public and private
+        // the public inputs should be at the beginning
+        func.sig.arguments.sort_by_key(|arg| match &arg.attribute {
+            Some(attr) => match attr.kind {
+                crate::parser::types::AttributeKind::Pub => 0,
+                _ => 1,
+            },
+            None => 1,
+        });
```
Comment (mimoo) [inline src/backends/r1cs/mod.rs:323]: imo this should happen outside of this function, unless it becomes a backend specific thing for now this code could leave outside

File: src/backends/r1cs/mod.rs [hunk 2]
```diff
@@ -374,6 +390,11 @@ where
 
         // check if every cell vars end up being a cell var in the circuit or public output
         for (index, _) in self.witness_vars.iter().enumerate() {
+            // skip the first var that is assumed to be the factor of constants of all linear combinations
```
Comment (mimoo) [inline src/backends/r1cs/mod.rs:393]: wdym "the factor of constants of all linear combinations"?

File: src/circuit_writer/mod.rs [hunk 3]
```diff
@@ -150,11 +150,13 @@ impl<B: Backend> CircuitWriter<B> {
         let qualified = FullyQualified::local("main".to_string());
         let main_fn_info = circuit_writer.main_info()?;
 
-        let function = match &main_fn_info.kind {
+        let mut function = match &main_fn_info.kind {
             crate::imports::FnKind::BuiltIn(_, _) => unreachable!(),
             crate::imports::FnKind::Native(fn_sig) => fn_sig.clone(),
         };
```
Comment (mimoo) [inline src/circuit_writer/mod.rs:156]: maybe clearer to write something like this:

```rust
// handle public inputs in order
for arg in function.sig.arguments.iter().filter(|arg| arg.is_public()) {
  handle_arg(arg, circuit_writer.add_public_inputs)?;
}

// handle private inputs in order
for arg in function.sig.arguments.iter().filter(|arg| arg.is_private())
  handle_arg(arg, circuit_writer.add_private_inputs)?;
}
```

which makes the handling of public inputs before private inputs much more explicit

File: examples/fixture/asm/r1cs/public_output.asm [hunk 4]
```diff
@@ -1,4 +1,4 @@
 @ noname.0.7.0
 
-2 == (v_1 + v_2) * (1)
-v_1 + v_2 + 6 == (v_3) * (1)
+2 == (v_2 + v_3) * (1)
+v_2 + v_3 + 6 == (v_1) * (1)
```
Comment (mimoo) [inline examples/fixture/asm/r1cs/public_output.asm:4]: this is really nice to read to review these kind of PRs :) looks correct to me

File: book/src/cellvar.md [hunk 5]
```diff
@@ -22,7 +22,7 @@ pub fn new_internal_var(&mut self, val: Value, span: Span) -> CellVar {
     self.next_variable += 1;
 
     // store it in the compiler
-    self.witness_vars.insert(var, val);
+    self.vars_to_value.insert(var, val);
```
Comment (mimoo) [inline book/src/cellvar.md:25]: :') ❤️

File: src/backends/mod.rs [hunk 6]
```diff
@@ -129,7 +129,7 @@ pub trait Backend: Clone {
 
     /// Compute the value of the symbolic cell variables.
     /// It recursively does the computation down the stream until it is not a symbolic variable.
-    /// - The symbolic variables are stored in the witness_vars.
+    /// - The cell vars are stored in the vars_to_value.
```
Comment (mimoo) [inline src/backends/mod.rs:132]: maybe remove that comment since this is a backend thing now

File: book/src/witness-generation.md [hunk 7]
```diff
@@ -16,17 +16,17 @@ pub struct Compiler {
     // ...
 
     /// This is used to compute the witness row by row.
-    pub rows_of_vars: Vec<Vec<Option<CellVar>>>,
+    pub witness_table: Vec<Vec<Option<CellVar>>>,
```
Comment (mimoo) [inline book/src/witness-generation.md:19]: ❤️

File: src/backends/r1cs/mod.rs [hunk 8]
```diff
@@ -215,7 +215,7 @@ where
 {
     /// Constraints in the r1cs.
     constraints: Vec<Constraint<F>>,
-    witness_vars: Vec<Value<Self>>,
+    witness_vector: Vec<Value<Self>>,
```
Comment (mimoo) [inline src/backends/r1cs/mod.rs:218]: this is awesome \o/

File: src/backends/r1cs/snarkjs.rs [hunk 9]
```diff
@@ -357,93 +294,3 @@ impl WitnessWriter {
...
-                    let new_index = snarkjs_exporter
-                        .witness_map
-                        .get(&original_var.index)
-                        .unwrap();
-                    assert!(rterms.contains_key(new_index));
-                }
-            }
-        }
-    }
-}
```
Comment (mimoo) [inline src/backends/r1cs/snarkjs.rs:449]: nice to see how much code we're removing through this PR :D

**Review by mimoo — APPROVED** [review]
woop woop!

File: src/backends/r1cs/mod.rs (hunk 2 above)
Comment (katat) [reply src/backends/r1cs/mod.rs:393]: The first var is assumed to be the coefficient of the constant of a linear combination.

In a linear combination, each of the vars can be paired with a coefficient.
The constant acts as the coefficient while var part is the always the first var (1). Look at it another way, the first var can always be the factor of the constants.

File: src/backends/r1cs/snarkjs.rs (hunk 9 above)
Comment (katat) [reply src/backends/r1cs/snarkjs.rs:449]: yeah, it is indeed simplify a lot things. Thanks for the suggestions!

File: src/circuit_writer/mod.rs (hunk 3 above)
Comment (katat) [reply src/circuit_writer/mod.rs:159]: Refactored: [99f328f](https://github.com/zksecurity/noname/pull/78/commits/99f328fa9d6799148eff5dd65a699758aeeceebe)

Thanks for the advice.

File: src/backends/r1cs/mod.rs (hunk 1 above)
Comment (katat) [reply src/backends/r1cs/mod.rs]: [99f328f](https://github.com/zksecurity/noname/pull/78/commits/99f328fa9d6799148eff5dd65a699758aeeceebe)

File: src/circuit_writer/mod.rs [hunk 10]
```diff
@@ -57,6 +57,9 @@ pub struct DebugInfo {
     pub note: String,
 }
 
+type HandleInput<B: Backend> =
```
Comment (mimoo) [inline src/circuit_writer/mod.rs:60]: (imo not worth creating a type that's only used in one place)

File: src/backends/r1cs/mod.rs (hunk 2 above)
Comment (mimoo) [reply src/backends/r1cs/mod.rs:383]: don't tell me, tell it in the comment :) I was just pointing out that the comment was not clear

**Pull request merged**